basic-4/splinepdf.py

```diff
@@ -4,17 +4,29 @@
 from scipy.interpolate import InterpolatedUnivariateSpline as IUS
 
 class ProbabilityDensityFunction(IUS):
-    def __init__(self, x, y):
+    """ Spline-based pdf class.
+    """
+    def __init__(self, x, y, k=3):
         if np.any(y<0):
             raise ValueError("Sampled pdf values must not be less than zero.")
-        not_normalized_pdf = IUS(x, y, ext="zeros")
-        norm_coeff = not_normalized_pdf.integral(-np.inf,np.inf)
-        IUS.__init__(self, x, y/norm_coeff, ext="zeros")
+        # Due to the fact that splines of order k have smoothness C^(k-1)
+        # some very sharp distribution may not be well described by a
+        # cubic spline. This means that the pdf may have values < 0,
+        # hence the cdf is not monotone, hence not invertible.
+        raw_spl_int = IUS(x, y, k=k, ext="zeros").antiderivative()
+        if np.any(np.diff(raw_spl_int(x))<0.0):
+            raise RuntimeError(("Spline interpolation returned non-monotone "
```

Comment:
Suggerisco di separare questo messaggio di errore in:

- Messaggio di errore da printare sullo stdout:
   ```suggestion
            raise RuntimeError(("Spline interpolation returned non-monotone "
                                            "cumulative density function."))
```

- Le indicazioni per come risolvere il problema secondo me si potrebbero dare all'interno di una documentazione più estesa, chiamabile dall'utente con un metodo apposito (?). Questo potrebbe comunque integrare alcune indicazioni immediate nel messaggio di errore.
```suggestion
            raise RuntimeError(("Spline interpolation returned non-monotone "
                "cumulative density function."
                "\nTry to lower the spline order or to increase the number of sampled "
                "point to avoid this problem."
                "\nLearn about some solutions with the ProbabilityDensityFunction method:"
                "\n\tProbabilityDensityFunction.not_monotone_cdf()"))
```

- Non accennerei in questo punto il problema dell'approccio con la spline.